pennylane/transforms/decorators.py

```diff
@@ -0,0 +1,388 @@
+                    param = op.parameters[0]
+                    qml.RX(x * param, wires=wires[1])
+                    qml.RY(y * qml.math.sqrt(param), wires=wires[1])
+                    qml.CZ(wires=[wires[1], wires[0]])
+                else:
+                    op.queue()
+
+        @my_transform(0.6, 0.1)
+        def qfunc(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+
+    Let's use this qfunc to create a QNode, so that we can execute it on a quantum
+    device:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(2.5))
+     0: ──H───────────────────╭Z──┤
+     1: ──RX(1.5)──RY(0.158)──╰C──┤
+
+    The not only is the transformed qfunc fully differentiable, but the qfunc transform
+    parameters *themselves* are differentiable:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        def ansatz(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+
+        @qml.qnode(dev)
+        def circuit(param, transform_weights):
+            qml.RX(0.1, wires=0)
+
+            # apply the transform to the ansatz
+            my_transform(*transform_weights)(ansatz)(param)
+
+            return qml.expval(qml.PauliZ(1))
+
+    We can print this QNode to show that the qfunc transform is taking place:
+
+    >>> x = np.array(0.5, requires_grad=True)
+    >>> y = np.array([0.1, 0.2], requires_grad=True)
+    >>> print(qml.draw(circuit)(x, y))
+     0: ──H────────────────────╭Z──┤
+     1: ──RX(0.05)──RY(0.141)──╰C──┤ ⟨Z⟩
+
+    Evaluating the QNode, as well as the derivative, with respect to the gate
+    parameter *and* the transform weights:
+
+    >>> circuit(x, y)
+    0.9887793925354269
+    >>> qml.grad(circuit)(x, y)
+    (array(-0.02485651), array([-0.02474011, -0.09954244]))
+    """
+
+    if not isinstance(tape_transform, single_tape_transform):
+        raise ValueError("Can only convert single tape transforms into qfunc transforms!")
+
+    @functools.wraps(tape_transform)
+    def make_qfunc_transform(*targs, **tkwargs):
+        def wrapper(fn):
+            @functools.wraps(fn)
+            def internal_wrapper(*args, **kwargs):
+                tape = make_tape(fn)(*args, **kwargs)
+                tape = tape_transform(tape, *targs, **tkwargs)
+                return tape.measurements
+
+            return internal_wrapper
+
+        return wrapper
+
+    return make_qfunc_transform
+
+
+def qnode_transform(tape_transform):
+    """Register a new QNode transform.
+
+    Args:
+        tape_transform (tape transform): the tape transform
+            to convert into the QNode transform.
+
+            Allowed tape transforms must be functions of the following form:
+
+            .. code-block:: python
+
+                def tape_transform(tape, *args, **kwargs):
+                    ...
+                    return tapes, processing_fn

```

Comment:
The returned values here conflict with the new definition of `tape_transform`, since those are only supposed to return only a single tape. Is the idea that this is something that could wrap one of those, but return possibly more tapes (and the function)?

Reply:
Yep. Essentially:

- qfunc transforms and single tape transforms are, to all intense and purposes, identical. So if you are doing something like compilation (tape->tape), it should be a qfunc transform.

- The _only reason_ to go further and do a QNode transform is:
  * You need to do something more powerful, as in the case here where the transform is tape->tapes with post-processing.
  * You need access to more information, such as the QNode device.

In fact, there are downsides to creating a QNode transform with just a tape->tape transform: you lose things like composability.

So... yeah. This was deliberate to an extent, so the documentation should better communicate this 🤔

Reply:
Oh okay, gotcha! Yeah we should definitely emphasize that :+1: